server/db/driver/pg/upgrades.go

```diff
@@ -0,0 +1,324 @@
+const dbVersion = 2
+
+// The number of upgrades defined MUST be equal to dbVersion.
+var upgrades = []func(db *sql.DB) error{
+	// v1 upgrade adds the schema_version column to the meta table, possibly
+	// creating the table if it was missing.
+	v1Upgrade,
+
+	// v2 upgrade creates epochs_report table, if it does not exist, and
+	// populates the table with partial historical data from the epochs and
+	// matches table. This includes match volumes, high/low/start/end rates, but
+	// does not include the booked volume statistics in the book_buys* and
+	// book_sells* columns since this data requires a book snapshot at the time
+	// of matching to generate.
+	v2Upgrade,
+}
+
+// v1Upgrade adds the schema_version column and removes the state_hash column
+// from the meta table.
+func v1Upgrade(db *sql.DB) error {
+	// Create the meta table with the v0 scheme. Even if the table does not
+	// exists, we should not create it fresh with the current scheme since one
+	// or more subsequent upgrades may alter the meta scheme.
+	metaV0Stmt := `CREATE TABLE IF NOT EXISTS %s (state_hash BYTEA)`
+	metaCreated, err := createTable(db, metaV0Stmt, publicSchema, metaTableName)
+	if err != nil {
+		return fmt.Errorf("failed to create meta table: %w", err)
+	}
+	if metaCreated {
+		log.Infof("Created new %q table", metaTableName)    // from 0.2+pre master
+		_, err = db.Exec(`INSERT INTO meta DEFAULT VALUES`) // might be CreateMetaRow, but ping to v0 stmt
+		if err != nil {
+			return fmt.Errorf("failed to create row for meta table: %w", err)
+		}
+	} else {
+		log.Infof("Existing %q table", metaTableName) // from release-0.1
+	}
+
+	// Create the schema_version column. The caller must set the version to 1.
+	_, err = db.Exec(`ALTER TABLE ` + metaTableName + ` ADD COLUMN IF NOT EXISTS schema_version INT4 DEFAULT 0;`)
+	if err != nil {
+		return err
+	}
+	_, err = db.Exec(`ALTER TABLE ` + metaTableName + ` DROP COLUMN IF EXISTS state_hash;`)
+	return err
+}
+
+// matchStatsForMarketEpoch is used by v2Upgrade to retrieve match rates and
+// quantities for a given epoch.
+func matchStatsForMarketEpoch(stmt *sql.Stmt, epochIdx, epochDur uint64) (rates, quantities []uint64, sell []bool, err error) {
+	var rows *sql.Rows
+	rows, err = stmt.Query(epochIdx, epochDur)
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var rate, quantity uint64
+		var takerSell bool
+		err = rows.Scan(&quantity, &rate, &takerSell)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		rates = append(rates, rate)
+		quantities = append(quantities, quantity)
+		sell = append(sell, takerSell)
+	}
+
+	if err = rows.Err(); err != nil {
+		return nil, nil, nil, err
+	}
+	return
+}
+
+// v2Upgrade populates the epoch_reports table with historical data from the
+// matches table.
+func v2Upgrade(db *sql.DB) error {
+	mkts, err := loadMarkets(db, marketsTableName)
+	if err != nil {
+		return fmt.Errorf("failed to read markets table: %w", err)
+	}
+	tx, err := db.Begin()
```

Comment:
Hmm, that could work for cancellation since the Rollback would catch it.  This upgrade is expected to take several minutes on busy servers, so I don't think a timeout would be easy to pick.  Let me wire up ctx so the operator can bail out.

Reply:
That was a minor rabbit hole, but I addressed this and making the upgrade + version bump atomic with a sql.Tx in https://github.com/decred/dcrdex/pull/959/commits/6b4d7c73fb65966372f31e5b7344e47e2cd37aa4.  It's very hard to write a test for it, but it did verify the functionality of a rollback I believe.  I also re-tested the massive upgrade of the DB referenced in the PR description.